### PR TITLE
[Merl 1256] Bug fix. Generate Autosave is missing from Post/Page Previews

### PR DESCRIPTION
**Pull request by theodesp**
## Tasks

- [ ] I have signed a [Contributor License Agreement (CLA)](https://github.com/wpengine/faustjs#contributor-license-agreement) with WP Engine.
- [ ] If a code change, I have written testing instructions that the whole team & outside contributors can understand.
- [ ] I have written and included a comprehensive [changeset](https://github.com/wpengine/faustjs/blob/canary/DEVELOPMENT.md#deployment) to properly document the changes I've made.

## Description

Fixes missing call to `autosave` when clicking on the Preview links.

## Testing

1. Checkout this branch that includes changes to the faustwp plugin
2. Using the updated faustwp plugin, create a new post & save as draft (do not publish)
3. Observe that both preview links correctly resolve to display draft content.
4. Verify that the draft content is also presented in the headless site.

### CODE REVIEW

File: plugins/faustwp/includes/replacement/previewlinks.js [hunk 1]
```diff
@@ -34,25 +32,51 @@ document.addEventListener('DOMContentLoaded', function() {
...
+      }
+    }
+  }
+
   function updateUIElements() {
     const { headerLink, snackbarLink } = getPreviewLinksByVersion(wpVersion);
 
-    // Clone & replace the original link in order to clear pre-existing events.
     if (headerLink && headerLink.getAttribute('href') !== faustPreviewLink) {
-      const clonedHeaderLink = headerLink.cloneNode(true);
```
Comment (theodesp) [inline plugins/faustwp/includes/replacement/previewlinks.js:48]: No DOM modifications allowed under React managed DOM

Screenshot 2023-11-13 at 13 32 12

**Comment by blakewilson** [comment]
@theodesp I've tested this and both existing previews and draft previews no longer work for me.

When requesting an existing post preview, the `post` comes back as `null`:
Screenshot 2023-11-13 at 8 32 02 AM

When requesting a new draft post, the `previewPathname` is no longer being added to the URL, which is causing an error on the Faust/Next side to be thrown. Oddly, when hovering over the preview button you can see in the bottom left corner, the URL does include the `previewPathname`:
Screenshot 2023-11-13 at 8 27 35 AM

But once the link is clicked, it is removed from the URL:
Screenshot 2023-11-13 at 8 27 51 AM

File: package.json [hunk 2]
```diff
@@ -80,6 +80,7 @@
   },
   "overrides": {
     "semver": "~7.5.2",
-    "optionator": "0.9.3"
+    "optionator": "0.9.3",
+    "axios": "^1.6.1"
```
[the remark this answers is not in the document]
Comment (theodesp) [reply package.json:84]: https://github.com/wpengine/faustjs/pull/1650

**Comment by theodesp** [comment]
@blakewilson  I've updated the code. It should work now:

https://github.com/wpengine/faustjs/assets/328805/c6fc76db-a589-4a6e-a87e-52aa69b7b500

**Comment by theodesp** [comment]
Lint error looks like relevant to this

https://github.com/woocommerce/woocommerce-sniffs/issues/34

Need to investigate

**Comment by blakewilson** [comment]
> @theodesp I've tested this and both existing previews and draft previews no longer work for me.
> 
> When requesting an existing post preview, the `post` comes back as `null`: Screenshot 2023-11-13 at 8 32 02 AM
> 
> When requesting a new draft post, the `previewPathname` is no longer being added to the URL, which is causing an error on the Faust/Next side to be thrown. Oddly, when hovering over the preview button you can see in the bottom left corner, the URL does include the `previewPathname`: Screenshot 2023-11-13 at 8 27 35 AM
> 
> But once the link is clicked, it is removed from the URL: Screenshot 2023-11-13 at 8 27 51 AM

I've realized this was due to a bug in Faust/Safari, so this may be unrelated.

**Review by blakewilson — APPROVED** [review]
Nice work @theodesp! I've tested locally for published posts and draft posts and both work as expected.

File: plugins/faustwp/includes/replacement/callbacks.php [hunk 3]
```diff
@@ -244,6 +244,24 @@ function enqueue_preview_scripts() {
 	);
 }
 
+add_filter( 'rest_prepare_post', __NAMESPACE__ . '\\preview_link_in_rest_response', 10, 2 );
+add_filter( 'rest_prepare_page', __NAMESPACE__ . '\\preview_link_in_rest_response', 10, 2 );
```
Comment (mindctrl) [inline plugins/faustwp/includes/replacement/callbacks.php:248]: Can we add some `has_action` tests for these? They seem like silly tests, but I've had them catch unintentional removals before, so I typically add them.

**Review by mindctrl — CHANGES_REQUESTED** [review]
Left comment about tests. Also there's a merge conflict now.

**Review by matthewguywright — APPROVED** [review]

**Comment by TeresaGobble** [comment]
Good afternoon Theo! I tried running this branch using the app-router example to play with it (it's connected to my test site via the `.env.local`, and I have the active plugin symlinked into my test site). I have received an error I haven't seen before when running `npm run dev`, below: 
image

I've already deleted `node_modules

Is there something amiss with the way I'm approaching this for testing?

**Comment by TeresaGobble** [comment]
Apologies for accidentally closing this, typing in github and my focus jumped to `close with comment`.

**Pull request merged**